### PR TITLE
7️⃣ V7

### LINKED ISSUE #1680: Support React Fast Refresh
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### LINKED ISSUE #2086: Rename dirtyFields to dirty or touched to touchedFields in formState for consistency
**Is your feature request related to a problem? Please describe.**
That's definitely a very minor thing, but perfectionist in me is bit frustrated by naming inconsistency for touched and dirty fields. 

**Describe the solution you'd like**
Rename dirtyFields to dirty or touched to touchedFields

### LINKED ISSUE #2155: [useFieldArray] customize autoFocus index
**Is your feature request related to a problem? Please describe.**

It would be nice developer can control  auto focus using `field name` or `field index`.

The use case is when the first field has default value, we may want to focus second/third input instead.

**Describe the solution you'd like**

`fieldArray.append(data, boolean | number | string )`

If second argument is boolean, same as before, focus first field

If second argument is number, focus field by index;
If second argument is string, focus field by name;  Maybe?

**Describe alternatives you've considered**

**Additional context**

https://github.com/react-hook-form/react-hook-form/blob/master/src/useFieldArray.ts#L436-L447

### LINKED ISSUE #2887: ESlint TypeScript error `unbound-method`
**Describe the bug**
Deconstructing `register` from `useForm` triggers `@typescript-eslint` error [`unbound-method`](https://github.com/typescript-eslint/typescript-eslint/blob/v4.1.0/packages/eslint-plugin/docs/rules/unbound-method.md)

**To Reproduce**
Use ESlint with a rule mentioned above.

**Codesandbox link (Required)**
CodeSandbox does not support ESlint (https://github.com/codesandbox/codesandbox-client/issues/1933)

**Expected behavior**
`register` is not a method that depends on `this` context, so should not trigger a warning. **`handleSubmit` doesn't**.

**Screenshots**
Screenshot from 2020-09-11 15-40-37

**Desktop (please complete the following information):**

- OS: btw, I use Arch (not really, Manjaro)
- Browser Vivaldi
- Version up to date
:P

**Additional context**
Similar problem with Formik: https://github.com/formium/formik/issues/2589

*PR welcome? I can contribute*

### LINKED ISSUE #3045: `unregister(...)` optional shouldDirty configuration parameter
**Is your feature request related to a problem? Please describe.**
I'm trying to unregister inputs and have them not affect the `isDirty` and `dirtyFields` properties.
Currently, after an input is unregistered, it sets `isDirty` to `true`.

**Describe the solution you'd like**
`unregisters` takes a second argument of type `{ shouldDirty: boolean }`.
This will be a non-braking change, and also provide future proof if more configs need to be added.

**Describe alternatives you've considered**
Having a `removeInput` function that removes all references to the input from the state.

### LINKED ISSUE #3110: Provide the resolver function with the name of the field that was changed
**Is your feature request related to a problem? Please describe.**
I have a form containing a field array and each item in it is a form (let's call them subforms) containing around 30 fields. I'd like use `onChange` mode for validation as it provides immediate feedback but this expectedly causes performance issues when the subforms are more than 5.

The general `Yup` schema goes something like this:

`
const schema = object().shape({
  subforms: array().of(getSubformSchema())
})`

**Describe the solution you'd like**
The validation for each subform mentioned above is actually independent of the others. So when `react-hook-form` calls the validation resolver it can pass the name of the field that was changed (this will be undefined if the resolver is not called as a result of a field change). Using this name we can potentially optimise how the form is validated if the validation schema library supports it. 

For example `Yup` provides the method `validateAt` which can validate a deeply nested schema. So when the resolver function is called with a field name we can use that name to find the parent subform of the field and only validate it, not the other subforms. 

In general, I think having access to the name of the field that was changed and caused the resolver to be called, could be useful for other cases too.  

**Describe alternatives you've considered**
Having multiple `useForm` calls, but this is hard as useForm is a hook and the count of the subforms is dynamic.

### LINKED ISSUE #3155: Unregister all input fields
**Is your feature request related to a problem? Please describe.**
Problem discussion: https://github.com/react-hook-form/react-hook-form/discussions/3153

**Describe the solution you'd like**
Have a `unregister()` method with empty params to unregister all the input fields.

### LINKED ISSUE #3538: PROPOSAL: Pass the name/ref of the field being validated over to resolvers
Hello 👋
Thank you for this awesome library! The amount of effort put into it is just incredible!

## Context
I am looking into writing custom resolver for [Vest](https://github.com/ealush/vest). Vest has a feature that allows the you to [only validate the field the user is currently interacting it](https://ealush.com/vest/#/./exclusion?id=only-running-specific-tests-including), while the rest of the results are cached internally (so no errors are lost when switching between fields).

This is quite a big perf-saver when working in an onChange or onBlur mode, because we do not need to worry about running everything and then filtering out the irrelevant errors.

The only thing required to allow this feature in HF as well is to pass the field (or fields) name (when applicable) being validated to the resolver. At the moment this behavior is only supported in `validateField` which does accept a reference to a field, but does not call the resolver.

The resolver would then be able to either use `vest.only` on behalf of the consumer, or pass it over for the consumers to use directly - depending on the validation strategy.

## Requested interface change
I would expect to use it like this:
```js
const { register, errors} = useForm({
	// fieldNames is either undefined, a string or an array of strings.
    resolver: (values, context, isValidateAllFieldCriteria, fieldNames) => {

      const result = validate(values, fieldNames);

      const errors = {};
      const vestErrors = result.getErrors();
      for (const field in vestErrors) {
        errors[field] = vestErrors[field][0];
      }

      return { values, errors };
    },
    mode: "onChange"
  })
```

### Alternative 1

One alternative would be to add an `extra`/`options` object. It would be effectively the same, but support future extensions of the interface without introducing bloat to the function signature.

```js
const { register, errors} = useForm({
    resolver: (values, context, isValidateAllFieldCriteria, extra = {}) => {

      const result = validate(values, extra.fieldNames);

      const errors = {};
      const vestErrors = result.getErrors();
      for (const field in vestErrors) {
        errors[field] = vestErrors[field][0];
      }

      return { values, errors };
    },
    mode: "onChange"
  })
```

### Alternative 2

Another viable alternative that does not require changing the function signature is extending the context ref with the current field/fields name. This might introduce a breaking change, since consumers may be using `context` to store primitives.
Also, it would require a post-validation cleanup of the context.

```js
// Using a symbol to prevent  a clash with consumer properties.
const INTERNAL_FIELD_NAMES = Symbol();
// ...
const { register, errors} = useForm({
    resolver: (values, context, isValidateAllFieldCriteria) => {

      const result = validate(values, context[INTERNAL_FIELD_NAMES]);

      const errors = {};
      const vestErrors = result.getErrors();
      for (const field in vestErrors) {
        errors[field] = vestErrors[field][0];
      }

      return { values, errors };
    },
    mode: "onChange"
  })
```

## Planned lib changes

I've mapped out the code changes I believe are needed in order to support this addition. There may be more that I missed. Let me know if you would like me to take a stab at it, or if you prefer a different approach.

#### Change 1 // for trigger()

https://github.com/react-hook-form/react-hook-form/blob/279061c7df264dd0549d050a294e57dc509b3447/src/useForm.ts#L382-L392
```js
const executeSchemaOrResolverValidation = React.useCallback(
  async (
    names:
      | InternalFieldName<TFieldValues>
      | InternalFieldName<TFieldValues>[],
  ) => {
    const { errors } = await resolverRef.current!(
      getValues(),
      contextRef.current,
      isValidateAllFieldCriteria,
	  names
    );
```

#### Change 2 // handleChangeRef

https://github.com/react-hook-form/react-hook-form/blob/279061c7df264dd0549d050a294e57dc509b3447/src/useForm.ts#L614-L618
```js
const { errors } = await resolverRef.current(
  getValues(),
  contextRef.current,
  isValidateAllFieldCriteria,
  name
);
```

#### Change 3 // validateResolver implementation
https://github.com/react-hook-form/react-hook-form/blob/279061c7df264dd0549d050a294e57dc509b3447/src/useForm.ts#L712-L721

```ts
const validateResolver = React.useCallback(
async (values = {}, fieldName?: string | string[]) => { 
  const { errors } = await resolverRef.current!( 
    { 
      ...getValues(), 
      ...values, 
    }, 
    contextRef.current, 
    isValidateAllFieldCriteria,
    fieldName
  ); 
```

#### Change 4 // validateResolver usage
https://github.com/react-hook-form/react-hook-form/blob/279061c7df264dd0549d050a294e57dc509b3447/src/useForm.ts#L776-L779

```js
readFormStateRef.current.isValid &&
  resolverRef.current &&
  validateResolver(undefined, field.ref.name);
updateWatchedValue(field.ref.name);
```

### LINKED ISSUE #3698: Seeking a way to watch for changes via `setValue`, since `<Form onChange/>` won't work
**Is your feature request related to a problem? Please describe.**
My current app has a number of forms with duplicated logic, so I have a `<CustomForm>` element I use to handle the logic. One of the most important parts is that `CustomForm` handles watching for changes, then sends them to an iframe to make live updates.

This works great for user-initiated changes to form fields, but it fails when I need to use `setValue`, naturally because it doesn't trigger a `change` event. One example is a rich text field (using `slate`). Another is a drag-and-drop avatar uploader.

So I'm seeking some way to watch for _all_ changes to the form data. `watch()` isn't a great solution because I need to know what field just changed, not just that the entire object changed. I would prefer not to do a deep equal for every character a user types, you know?

**Describe the solution you'd like**

Some ideas:
* Maybe `setValue` triggers a synthetic change event.
* Maybe `watch()` receives a callback function `cb(name, value, ref)`.

**Additional context**
I've looked through the issues and discussion. It seems like this is a fairly common pain point. That said, I can definitely imagine I'm just missing an easy way to do this as is.

### PR DESCRIPTION
**Pull request by bluebill1049**
## Vision

- 📖 (DX) Strict typed form
- 🏋🏻‍♀️ Reduce package size
- 🏎 Performance enhancements
- 💁‍♂️ improve API's simplicity and consistency

RFC: https://github.com/react-hook-form/react-hook-form/discussions/3714

-----

### API

useForm:

- [x] register
- [x] unregister
- [x] reset
- [x] watch
- [x] trigger
- [x] errors
- [x] touchedFields

useFieldArray
- [x] append
- [x] insert
- [x] prepend

Controller
- [x] Controller

useFormState
- [x] useFormState

### Resolver 

- [x] resolver (core)
- [x] @hookform/resolver

### DevTool

- [x] @hookform/devtool https://github.com/react-hook-form/devtools/pull/66

### Docs

- [x] documentation  https://github.com/react-hook-form/documentation/pull/510
- [x] examples

### Features/Issues

- close #2887
- close #2086
- close #2155
- close #3538
- close #3110
- close #3698
- close #1680
- close #3155
- close #3045

### CODE REVIEW

**Review by mestir — APPROVED** [review]

**Review by mestir — DISMISSED** [review]
Leave

**Comment by mestir** [comment]
Bonjour]()

**Review by mestir — DISMISSED** [review]

**Review by alfredguedea — DISMISSED** [review]

**Comment by kotarella1110** [comment]
I will improve all useForm methods to be type-safe:

- [ ] trigger
- [ ] setValue
- [ ] clearErrors
- [ ] setErrors

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,284 @@
...
+```diff
+- const { field, meta } = useController({ control });
++ const { field, fieldState, formState } = useController({ control });
+```
+
+## [7.0.0-beta.9] - 2021-02-19
+
+## Changes
+
+- typescript array index support is changed to `49` instead of `99`
```
Comment (pmaier983) [inline CHANGELOG.md:49]: Could someone link me to docs or the reason for this? just curious 😄

**Review by pmaier983 — COMMENTED** [review]
Hey bill, I'll be going through this PR in the next day or two. How do you want me to review it? Because of the size of this PR I think it might be easier to open a PR into this PR, or I can keep adding suggestions, whatever is easiest for you! 

P.S. Any areas you want me to target?

Best,
Phillip

File: CHANGELOG.md (hunk 1 above)
Comment (bluebill1049) [reply CHANGELOG.md:49]: no longer the case :) @leapful solved it 👍

**Comment by bluebill1049** [comment]
> Hey bill, I'll be going through this PR in the next day or two. How do you want me to review it? Because of the size of this PR I think it might be easier to open a PR into this PR, or I can keep adding suggestions, whatever is easiest for you!
> 
> P.S. Any areas you want me to target?
> 
> Best,
> Phillip

Thanks, @pmaier983, feel free to scan through the change and leave questions or sending PR for better solutions. There is no specific area to be focus rather look different parts and see if any issue or improvements.

File: examples/V7/validationSchema.tsx [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    validationSchema: SignupSchema,
+  });
+  const onSubmit = (data) => {
+    alert(JSON.stringify(data));
+  };
+
+  return (
+    <form onSubmit={handleSubmit(onSubmit)}>
+      <div>
+        <label>Develop?</label>
```
Comment (pmaier983) [inline examples/V7/validationSchema.tsx:23]:
```suggestion
        <label>Are you a Developer?</label>
```

File: examples/V7/validationSchema.tsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        Yes
+        <input type="radio" {...register('test')} value="yes" />
+        No
+        <input type="radio" {...register('test')} value="no" />
```
Comment (pmaier983) [inline examples/V7/validationSchema.tsx:27]: Should these have the same register key as they are radio values? is there some functionality I am unaware of?

**Review by pmaier983 — COMMENTED** [review]
Got through the example section. More to come 😄

File: examples/V7/basicValidation.tsx [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+      </div>
+      <div>
+        <label>Title</label>
+        <select {...register('Title', { required: true })}>
+          <option value="Mr">Mr</option>
+          <option value="Mrs">Mrs</option>
+          <option value="Miss">Miss</option>
+          <option value="Dr">Dr</option>
+        </select>
+      </div>
```
[the remark this answers is not in the document]
Comment (bluebill1049) [reply examples/V7/basicValidation.tsx:57]: yep, agree 👍

File: examples/V7/customValidation.tsx [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+export default function App() {
+  const { register, handleSubmit, errors } = useForm();
+  const onSubmit = (data) => {
+    alert(JSON.stringify(data));
+  };
+  const intialValues = {
+    firstName: 'bill',
+    lastName: 'luo',
+    email: 'bluebill1049@hotmail.com',
+    age: -1,
```
[the remark this answers is not in the document]
Comment (bluebill1049) [reply examples/V7/customValidation.tsx:14]: 👍

File: examples/V7/validationSchema.tsx (hunk 3 above)
Comment (bluebill1049) [reply examples/V7/validationSchema.tsx:27]: yes, for radio input with same name and different value.

File: examples/V7/initalFormState.tsx [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+const defaultValues = {
+  firstName: 'bill',
+  lastName: 'luo',
+  email: 'bluebill1049@hotmail.com',
+};
+
+function App() {
+  const { register, handleSubmit } = useForm();
+  // or you can set up the defaultValues at useForm
```
[the remark this answers is not in the document]
Comment (bluebill1049) [reply examples/V7/initalFormState.tsx:15]: 👍

**Comment by bluebill1049** [comment]
thanks heaps @pmaier983 for all the corrections ❤️

**Pull request merged**